cmd/lotus-chainwatch/sync.go

```diff
@@ -263,57 +305,97 @@ func syncHead(ctx context.Context, api api.FullNode, st *storage, headTs *types.
 
 		log.Infof("Getting miner info")
 
-		miners := map[minerKey]*minerInfo{}
+		// map of tipset to all miners that had a head-change at that tipset.
+		minerTips := map[types.TipSetKey][]*newMinerInfo{}
+		// heads we've seen, im being paranoid
+		headsSeen := map[cid.Cid]bool{}
```

Comment:
I don't think it's going to make much of a difference, but because you are only doing an existence check here, you can make this a `map[cid.Cid]struct{}{}`

Fairly good example: https://github.com/ipfs/go-ipfs/blob/7ce1d751f808d6a990dc496a81f2477742f9e640/core/commands/bootstrap.go#L312-L338